### PR TITLE
t/131: Contextual toolbar sample

### PR DESCRIPTION
**Pull request by oleq**
It's a demo of #131.

Depends on #98 and https://github.com/ckeditor/ckeditor5-theme-lark/pull/55.

### CODE REVIEW

File: tests/manual/createcontextualtoolbar/createcontextualtoolbar.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import BalloonPanelView from 'ckeditor5/ui/balloonpanel/balloonpanelview.js';
+
+const positions = {
+	//	     [ Target ]
+	//	              ^
+	//	     +-----------------+
+	//	     |     Balloon     |
+	//	     +-----------------+
+	forwardSelection: ( targetRect, balloonRect ) => ( {
+		top: targetRect.bottom + 15,
```
Comment (Reinmar) [inline tests/manual/createcontextualtoolbar/createcontextualtoolbar.js]: What's 15?

Comment (oleq) [reply tests/manual/createcontextualtoolbar/createcontextualtoolbar.js]: https://github.com/ckeditor/ckeditor5-ui-default/blob/t/131/src/balloonpanel/balloonpanelview.js#L19 I forgot to make these public. Static class properties?

File: src/balloonpanel/balloonpanelview.js [hunk 2]
```diff
@@ -159,6 +159,43 @@ export default class BalloonPanelView extends View {
...
+ *		 |     |
+ *		 +--+  |  +------...
+ *		     \ | /
+ *		      \|/
+ *	    >|-----|<---------------- horizontal offset
+ *
+ * @default 30
+ * @member {Number} module:ui/balloonpanel/balloonpanelview~BalloonPanelView.arrowHorizontalOffset
+ */
+BalloonPanelView.arrowHorizontalOffset = arrowHorizontalOffset;
```
Comment (Reinmar) [inline src/balloonpanel/balloonpanelview.js]: Why are you exposing these values if you're not reading it later on but use the consts? This means that if someone changes the theme and try to change these values it's not going to work.

File: tests/manual/contextualtoolbar/contextualtoolbar.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+const positions = {
+	//     [text range]
+	//                ^
+	//       +-----------------+
+	//       |     Balloon     |
+	//       +-----------------+
+	forwardSelection: ( targetRect, balloonRect ) => ( {
+		top: targetRect.bottom + arrowVOffset,
+		left: targetRect.right - balloonRect.width / 2,
+		name: 's'
```
Comment (Reinmar) [inline tests/manual/contextualtoolbar/contextualtoolbar.js:31]: I'm thinking now... how does this value affect anything? It certainly isn't used anywhere in this code. So perhaps it doesn't need to be defined at all? Am I right that only a feature which wants to understand which position the `getOptimalPosition` function had returned  needs to define it?

Comment (Reinmar) [reply tests/manual/contextualtoolbar/contextualtoolbar.js:31]: OK, I see that these values are also used in themes. So you should also update `BallonPanelView#position` values. Or even better – extract it to some typedef and document that it's related to the theme's classes.

**Review by Reinmar — CHANGES_REQUESTED** [review]
It doesn't work on Firefox at all. Apparently there's either no click event when selection is made or click is fired before the selection is made.

**Comment by Reinmar** [comment]
The same problem can be reproduced in Safari.

File: tests/manual/contextualtoolbar/contextualtoolbar.js (hunk 3 above)
Comment (oleq) [reply tests/manual/contextualtoolbar/contextualtoolbar.js:31]: I'm not quite sure what you expect me to change. Positions are documented https://github.com/ckeditor/ckeditor5-utils/blob/master/src/dom/position.js#L263-L272. A position must have a name for any code using the `getOptimalPosition` utility to know, which of the positions have been chosen, not only what coordinates represent it. We've discussed it already.

**Comment by oleq** [comment]
As for the Safari, the editor won't start in this browser and it has nothing to do with the scope of the PR. The question is: where's an issue for that?

**Comment by Reinmar** [comment]
> As for the Safari, the editor won't start in this browser and it has nothing to do with the scope of the PR. The question is: where's an issue for that?

Update ckeditor5-dev-tests. It has been fixed.

File: tests/manual/contextualtoolbar/contextualtoolbar.js (hunk 3 above)
Comment (Reinmar) [reply tests/manual/contextualtoolbar/contextualtoolbar.js:31]: What I'm expecting is that the set of supported names is defined somewhere. Initially I thought that this set is totally open, but it's not – the names must much the positions supported by the theme. So we should indicate (to help the developer understanding what it all is) that this is a set of 6 values *and* that you can add more if you style them.

**Comment by oleq** [comment]
> Update ckeditor5-dev-tests. It has been fixed.

No improvement.

image

**Comment by Reinmar** [comment]
That's exactly the issue I fixed recently and it WFM:

image


Make sure to update ckeditor5 before and then reinstall ckeditor5-dev-tests.

**Comment by Reinmar** [comment]
That was the issue: https://github.com/ckeditor/ckeditor5-dev-tests/issues/40. And the code landed in 3.1.1.

**Comment by oleq** [comment]
...and hasn't been updated on master

```
(master 1ea8f69) oleq@Aleksanders-MBP.local ~/CK/5/ckeditor5> grep ckeditor5-dev-tests package.json
    "@ckeditor/ckeditor5-dev-tests": "^3.1.0",
```

**Comment by Reinmar** [comment]
> ...and hasn't been updated on master

^3.1.0 satisfies 3.1.1. So everything was up to date.

**Comment by oleq** [comment]
So no idea then. Updated all packages, `npm i` and there's still that error.

**Comment by Reinmar** [comment]
> So no idea then. Updated all packages, npm i and there's still that error.

I know – that's why `npm i` is useless. And `npm update` is unreliable, so I prefer to remove the packages which I want to update and just install them again.

**Comment by oleq** [comment]
OK. Now I got it. It works in Safari. 

But the whole npm thing is weird, really. The problem will return sooner or later.

**Comment by Reinmar** [comment]
> But the whole npm thing is weird, really. The problem will return sooner or later.

It returns every day. It's how npm works and either we learn how to live with it or quit or jobs ;P We can also consider Yarn, but it didn't work in our case previously.

**Comment by Reinmar** [comment]
Toolbar should disappear when selection changes:

dec-16-2016 14-43-31

**Comment by oleq** [comment]
> Toolbar should disappear when selection changes:

This is but a sample, an entry point to actual implementation. There are still many behaviors that will need to be implemented, and this is just one of them. The sample should show the idea as simple as possible; it's up to actual integration to make it perfect. So I don't think this belongs to the scope of this PR.

**Comment by Reinmar** [comment]
Hm... OK. Let's polish the behaviour in actual usage places.

**Pull request merged**